File: packages/python/src/kweaver/resources/resources.py
"""SDK resource: vega-backend Resources (table/logicview)."""

from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING, Any
from urllib.parse import quote

logger = logging.getLogger(__name__)

from kweaver.types import Resource, ViewField

_DEFAULT_LIST_LIMIT = 30

if TYPE_CHECKING:
    from kweaver._http import HttpClient


class ResourcesResource:
    """Vega-backend resource management: table/logicview resources under a catalog."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def find_by_table(
        self,
        datasource_id: str,
        table_name: str,
        *,
        wait: bool = True,
        timeout: float = 30,
    ) -> Resource | None:
        """Find the resource for a specific table in a catalog (datasource)."""
        deadline = time.monotonic() + timeout
        attempt = 0
        while True:
            data = self._http.get(
                "/api/vega-backend/v1/resources",
                params={"catalog_id": datasource_id, "name": table_name},
            )
            items = data if isinstance(data, list) else (data.get("entries") or data.get("data") or [])
            logger.debug(
                "find_by_table attempt=%d ds=%s table=%r found=%d",
                attempt + 1, datasource_id, table_name, len(items),
            )
            for d in items:
                if d.get("name") == table_name:
                    return _parse_resource(d)
            if not wait or time.monotonic() >= deadline:
                return None
            delay = min(5.0, 1.0 * 2 ** attempt)
            time.sleep(delay)
            attempt += 1

    def create(
        self,
        name: str,
        datasource_id: str,
        *,
        table: str | None = None,
        category: str = "table",
        fields: list[dict[str, Any]] | None = None,
    ) -> Resource:
        """Create a vega-backend resource.

        For table-based resources (*table* is provided), posts category=table.
        For logic views, set *category='logicview'* and omit *table*.
        """
        body: dict[str, Any] = {
            "name": name,
            "catalog_id": datasource_id,
            "category": category,
        }
        if table:
            body["source_identifier"] = table
        if fields:
            body["schema_definition"] = fields
        data = self._http.post("/api/vega-backend/v1/resources", json=body)
        created_id = str(data.get("id", "")) if isinstance(data, dict) else ""
        if not created_id:
            raise ValueError(f"No id in create response: {data}")
        return self.get(created_id)

    def list(
        self,
        *,
        datasource_id: str | None = None,
        name: str | None = None,
        category: str | None = None,
        limit: int = _DEFAULT_LIST_LIMIT,
    ) -> list[Resource]:
        params: dict[str, Any] = {"limit": limit}
        if datasource_id:
            params["catalog_id"] = datasource_id
        if name:
            params["name"] = name
        if category:
            params["category"] = category
        data = self._http.get("/api/vega-backend/v1/resources", params=params)
        items = data if isinstance(data, list) else (data.get("entries") or data.get("data") or [])
        return [_parse_resource(d) for d in items]

    def get(self, id: str) -> Resource:
        data = self._http.get(f"/api/vega-backend/v1/resources/{quote(str(id), safe='')}")
        items = data if isinstance(data, list) else (data.get("entries") or [])
        raw = items[0] if items else (data if isinstance(data, dict) and "id" in data else None)
        if not raw:
            raise ValueError(f"Resource not found: {id}")
        return _parse_resource(raw)

    def delete(self, id: str) -> None:
        self._http.delete(f"/api/vega-backend/v1/resources/{quote(str(id), safe='')}")

    def query(
        self,
        id: str,
        *,
        offset: int = 0,
        limit: int = 50,
        need_total: bool = False,
        filter_condition: Any | None = None,
        sort: str | None = None,
        direction: str | None = None,
    ) -> dict[str, Any]:
        """Query data from a vega-backend resource.

        POST ``/api/vega-backend/v1/resources/{id}/data``.
        """
        body: dict[str, Any] = {
            "offset": offset,
            "limit": limit,
            "need_total": need_total,
        }
        if filter_condition is not None:
            body["filter_condition"] = filter_condition
        if sort is not None:
            body["sort"] = sort
        if direction is not None:
            body["direction"] = direction
        path = f"/api/vega-backend/v1/resources/{quote(str(id), safe='')}/data"
        data = self._http.post(
            path,
            json=body,
            headers={"x-http-method-override": "GET"},
        )
        return data if isinstance(data, dict) else {}


def _parse_resource(d: dict[str, Any]) -> Resource:
    raw_schema = d.get("schema_definition")
    schema: list[ViewField] | None = None
    if raw_schema:
        schema = [
            ViewField(
                name=f["name"],
                type=f.get("type", "varchar"),
                display_name=f.get("display_name"),
                comment=f.get("comment"),
            )
            for f in raw_schema
        ]
    return Resource(
        id=str(d.get("id", "")),
        name=d.get("name", ""),
        catalog_id=str(d.get("catalog_id", "")),
        category=d.get("category", ""),
        source_identifier=d.get("source_identifier"),
        status=d.get("status"),
        schema_definition=schema,
        logic_definition=d.get("logic_definition"),
    )

File: packages/typescript/src/api/resources.ts
import { HttpError } from "../utils/http.js";
import { buildHeaders } from "./headers.js";

export const RESOURCE_LIST_DEFAULT_LIMIT = 30;

/** Field metadata for a resource schema. */
export interface ViewField {
  name: string;
  type: string;
  display_name?: string;
  comment?: string;
}

/** Normalized vega-backend Resource model. */
export interface Resource {
  id: string;
  name: string;
  catalog_id: string;
  category: string;
  source_identifier?: string;
  status?: string;
  schema_definition?: ViewField[];
  logic_definition?: unknown;
}

export function parseResource(raw: Record<string, unknown>): Resource {
  const res: Resource = {
    id: String(raw.id ?? ""),
    name: String(raw.name ?? ""),
    catalog_id: String(raw.catalog_id ?? ""),
    category: String(raw.category ?? ""),
  };
  if (raw.source_identifier != null) res.source_identifier = String(raw.source_identifier);
  if (raw.status != null) res.status = String(raw.status);
  if (Array.isArray(raw.schema_definition) && raw.schema_definition.length > 0) {
    res.schema_definition = (raw.schema_definition as Record<string, unknown>[]).map((f) => ({
      name: String(f.name ?? ""),
      type: String(f.type ?? "varchar"),
      display_name: f.display_name != null ? String(f.display_name) : undefined,
      comment: f.comment != null ? String(f.comment) : undefined,
    }));
  }
  if (raw.logic_definition !== undefined) res.logic_definition = raw.logic_definition;
  return res;
}

function extractListPayload(data: unknown): unknown[] {
  if (Array.isArray(data)) return data;
  if (data && typeof data === "object") {
    const obj = data as Record<string, unknown>;
    const items = obj.entries ?? obj.data;
    if (Array.isArray(items)) return items;
  }
  return [];
}

export interface CreateResourceOptions {
  baseUrl: string;
  accessToken: string;
  name: string;
  datasourceId: string;
  table: string;
  fields?: Array<{ name: string; type: string }>;
  businessDomain?: string;
}

export async function createResource(options: CreateResourceOptions): Promise<string> {
  const {
    baseUrl,
    accessToken,
    name,
    datasourceId,
    table,
    fields = [],
    businessDomain = "bd_public",
  } = options;

  const base = baseUrl.replace(/\/+$/, "");
  const url = `${base}/api/vega-backend/v1/resources`;

  const body: Record<string, unknown> = {
    name,
    catalog_id: datasourceId,
    category: "table",
    source_identifier: table,
  };
  if (fields.length > 0) body.schema_definition = fields;

  const response = await fetch(url, {
    method: "POST",
    headers: {
      ...buildHeaders(accessToken, businessDomain),
      "content-type": "application/json",
    },
    body: JSON.stringify(body),
  });

  const responseBody = await response.text();
  if (!response.ok) {
    throw new HttpError(response.status, response.statusText, responseBody);
  }

  const data = JSON.parse(responseBody) as Record<string, unknown>;
  return String(data.id ?? "");
}

export interface ListResourcesOptions {
  baseUrl: string;
  accessToken: string;
  businessDomain?: string;
  /** Filter by catalog (data source) id. */
  datasourceId?: string;
  /** Server-side name filter. */
  name?: string;
  /** Category filter (e.g. "table", "logicview"). */
  category?: string;
  /** Max items; ignored when <= 0. */
  limit?: number;
}

export async function listResources(options: ListResourcesOptions): Promise<Resource[]> {
  const {
    baseUrl,
    accessToken,
    businessDomain = "bd_public",
    datasourceId,
    name,
    category,
    limit = RESOURCE_LIST_DEFAULT_LIMIT,
  } = options;

  const base = baseUrl.replace(/\/+$/, "");
  const url = new URL(`${base}/api/vega-backend/v1/resources`);
  if (datasourceId) url.searchParams.set("catalog_id", datasourceId);
  if (name) url.searchParams.set("name", name);
  if (category) url.searchParams.set("category", category);
  if (limit && limit > 0) url.searchParams.set("limit", String(limit));

  const response = await fetch(url.toString(), {
    method: "GET",
    headers: buildHeaders(accessToken, businessDomain),
  });

  const bodyText = await response.text();
  if (!response.ok) throw new HttpError(response.status, response.statusText, bodyText);

  const parsed = JSON.parse(bodyText) as unknown;
  const items = extractListPayload(parsed);
  return items
    .filter((item): item is Record<string, unknown> => item != null && typeof item === "object")
    .map(parseResource);
}

export interface DeleteResourceOptions {
  baseUrl: string;
  accessToken: string;
  id: string;
  businessDomain?: string;
}

export async function deleteResource(options: DeleteResourceOptions): Promise<void> {
  const { baseUrl, accessToken, id, businessDomain = "bd_public" } = options;
  const base = baseUrl.replace(/\/+$/, "");
  const url = `${base}/api/vega-backend/v1/resources/${encodeURIComponent(id)}`;
  const response = await fetch(url, {
    method: "DELETE",
    headers: buildHeaders(accessToken, businessDomain),
  });
  const bodyText = await response.text();
  if (!response.ok) throw new HttpError(response.status, response.statusText, bodyText);
}

export interface GetResourceOptions {
  baseUrl: string;
  accessToken: string;
  id: string;
  businessDomain?: string;
}

export async function getResource(options: GetResourceOptions): Promise<Resource> {
  const { baseUrl, accessToken, id, businessDomain = "bd_public" } = options;
  const base = baseUrl.replace(/\/+$/, "");
  const url = `${base}/api/vega-backend/v1/resources/${encodeURIComponent(id)}`;

  const response = await fetch(url, {
    method: "GET",
    headers: buildHeaders(accessToken, businessDomain),
  });

  const body = await response.text();
  if (!response.ok) throw new HttpError(response.status, response.statusText, body);

  const parsed = JSON.parse(body) as unknown;
  // vega-backend GET /:id returns { entries: [...] } (supports comma-separated ids)
  const items = extractListPayload(parsed);
  const raw = items.length > 0 ? items[0] : (parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : null);
  if (!raw || typeof raw !== "object") throw new HttpError(500, "Invalid response", body);
  return parseResource(raw as Record<string, unknown>);
}

export interface FindResourceOptions {
  baseUrl: string;
  accessToken: string;
  businessDomain?: string;
  /** Resource name to search for. */
  name: string;
  /** Filter by catalog (data source) id. */
  datasourceId?: string;
  /** When true, apply client-side exact name match after server filter (default false). */
  exact?: boolean;
  /** When true, poll until a result appears or timeout (default false). */
  wait?: boolean;
  /** Total wait budget in ms (default 30000). Only used when wait is true. */
  timeoutMs?: number;
}

function sleepMs(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function findResource(options: FindResourceOptions): Promise<Resource[]> {
  const {
    baseUrl,
    accessToken,
    businessDomain = "bd_public",
    name,
    datasourceId,
    exact = false,
    wait = false,
    timeoutMs = 30_000,
  } = options;

  const deadline = Date.now() + timeoutMs;
  let attempt = 0;

  while (true) {
    const list = await listResources({ baseUrl, accessToken, businessDomain, datasourceId, name });
    const results = exact ? list.filter((v) => v.name === name) : list;
    if (results.length > 0 || !wait || Date.now() >= deadline) return results;
    const delayMs = Math.min(5000, 1000 * 2 ** attempt);
    attempt += 1;
    await sleepMs(delayMs);
  }
}

export interface QueryResourceOptions {
  baseUrl: string;
  accessToken: string;
  id: string;
  offset?: number;
  limit?: number;
  needTotal?: boolean;
  filterCondition?: unknown;
  sort?: string;
  direction?: "asc" | "desc";
  businessDomain?: string;
}

export interface ResourceQueryResult {
  entries?: unknown;
  total_count?: number;
}

export async function queryResource(options: QueryResourceOptions): Promise<ResourceQueryResult> {
  const {
    baseUrl,
    accessToken,
    id,
    offset = 0,
    limit = 50,
    needTotal = false,
    filterCondition,
    sort,
    direction,
    businessDomain = "bd_public",
  } = options;

  const base = baseUrl.replace(/\/+$/, "");
  const url = `${base}/api/vega-backend/v1/resources/${encodeURIComponent(id)}/data`;

  const body: Record<string, unknown> = { offset, limit, need_total: needTotal };
  if (filterCondition !== undefined) body.filter_condition = filterCondition;
  if (sort !== undefined) body.sort = sort;
  if (direction !== undefined) body.direction = direction;

  const response = await fetch(url, {
    method: "POST",
    headers: {
      ...buildHeaders(accessToken, businessDomain),
      "content-type": "application/json",
      "x-http-method-override": "GET",
    },
    body: JSON.stringify(body),
  });

  const bodyText = await response.text();
  if (!response.ok) throw new HttpError(response.status, response.statusText, bodyText);

  const parsed = JSON.parse(bodyText) as unknown;
  if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
    return parsed as ResourceQueryResult;
  }
  return {};
}
